receiver/filereceiver/watcher.go

```diff
@@ -0,0 +1,135 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filereceiver"
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"regexp"
+
+	"github.com/fsnotify/fsnotify"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/otlp"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+var (
+	tracesUnmarshaler  = otlp.NewJSONTracesUnmarshaler()
+	logsUnmarshaler    = otlp.NewJSONLogsUnmarshaler()
+	metricsUnmarshaler = otlp.NewJSONMetricsUnmarshaler()
+)
+
+type watcher struct {
+	logger          *zap.Logger
+	path            string
+	watcher         *fsnotify.Watcher
+	logsConsumer    consumer.Logs
+	tracesConsumer  consumer.Traces
+	metricsConsumer consumer.Metrics
+}
+
+var (
+	logsFingerPrint    = regexp.MustCompile("\\{\\s*\"resourceLogs\"")
+	metricsFingerPrint = regexp.MustCompile("\\{\\s*\"resourceMetrics\"")
+	tracesFingerPrint  = regexp.MustCompile("\\{\\s*\"resourceSpans\"")
+)
+
+func (w *watcher) start() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	w.watcher = watcher
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				path := event.Name
+				if event.Op == fsnotify.Write || event.Op == fsnotify.Create {
+					err := w.processFile(path)
+					w.logger.Error("error reading", zap.String("path", w.path), zap.Error(err))
+				}
+
+			case err := <-watcher.Errors:
+				w.logger.Error("error watching ", zap.String("path", w.path), zap.Error(err))
+			}
+		}
+	}()
+
+	if err := watcher.Add(w.path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *watcher) processFile(path string) error {
+	fileHandle, _ := os.Open(path)
+	defer fileHandle.Close()
+	fileScanner := bufio.NewScanner(fileHandle)
+
+	var allErrors []error
+	for fileScanner.Scan() {
+		if err := fileScanner.Err(); err != nil {
+			allErrors = append(allErrors, err)
+			break
+		}
+		line := fileScanner.Bytes()
+		if len(line) < 12 {
+			continue
+		}
+		// {"resourceSpans"
+		if tracesFingerPrint.Match(line) && w.tracesConsumer != nil {
```

Comment:
I would decouple these conditions.  If the regex matches then the line represents a trace regardless whether a consumer is configured and it is unnecessary to check for other fingerprints.  The same applies to all signal types.

Reply:
good call, done